### PR TITLE
Overview resources untapped

### PR DESCRIPTION
**Pull request by rh-github-2015**
A row showing how many owned tiles contain unimproved luxury or strategic resources (or having the wrong improvement) - in other words 'to-do''s. Does only help finding them a treensy tiny little bit, but can still save time.

I'm not completely convinced this is perfect, as the other rows operate on a different scale, but still...

And sorry I didn't add the empty rows to the other translation files, I assume there's a tool for that.

image

### CODE REVIEW

**Comment by JackRainy** [comment]
> val sum = resourceDrilldown.filter { it.resource==resource && it.amount > 0 }.sumBy { it.amount }       // negatives used for untapped tiles

The negatives are used also for trading. If you take a look at your screenshot you will see that silver's total is counted wrong!

> And sorry I didn't add the empty rows to the other translation files, I assume there's a tool for that.

Yes, you need to add an empty row to the `template.properties` file and others will be generated.

**Comment by rh-github-2015** [comment]
Good catch. That's Facepalm® level.

**Comment by rh-github-2015** [comment]
image
Added benefit: Columns reporting _only_ missed chanced now have empty sum cells. Nicer.

File: core/src/com/unciv/logic/city/CityInfo.kt [hunk 1]
```diff
@@ -171,6 +171,20 @@ class CityInfo {
...
+    fun getCityUntappedResources(): ResourceSupplyList {
+        val cityResources = ResourceSupplyList()
+
+        // count missed chances: resource is known, tile owned, but not improved or improvement not matching / not active
+        for (tileInfo in getTiles().filter { it.resource != null }) {
+            val resource = tileInfo.getTileResource()
+            if (resource.resourceType==ResourceType.Bonus) continue
+            if (resource.revealedBy!=null && !civInfo.tech.isResearched(resource.revealedBy!!)) continue
+            if (getTileResourceAmount(tileInfo)==0)
+                cityResources.add(resource, -1, "Untapped")
```
Comment (JackRainy) [inline core/src/com/unciv/logic/city/CityInfo.kt]: It still confuses me, why it is counted as a negative.
"I have -5 untapped resources." - that sounds strange.
I would suggest to display them as a positive number in braces to indicate that it is not counted to the total.

For example:
Screenshot 2020-03-20 08 11 36

Comment (rh-github-2015) [reply core/src/com/unciv/logic/city/CityInfo.kt]: You're right. The technical negative is left over from abusing it as a flag, the display was conscious but I did hesitate. (n) sounds good. I fear commiting that may pose surprises as I merged 3.6.9 into my masters, and that branch is based on 3.6.8. I'll try in a few hours.

**Comment by rh-github-2015** [comment]
The idea to make them clickable seemed beyond me. The clickability of the promotions in  #2187 was trivial copy&paste (and even there I had misgivings as I do not know enough about the presentation layer to predict whether such a breaking out of a nested screen directly back to the caller would not create a memory leak), and this seemed more complicated.

But maybe... How about every tap generating (n) "Iron was discovered near Rome" notifications each with a different location? Then popping a simple message box with an explanation only once (how is a simple messagebox even done in this framework)? Or add the explanation as static text right below the resources table? The latter I think. 

The initial trigger to build this was the fact that the "tech researched" trigger would always give you at most one such message and be gone next turn. Maybe I could also find that and rework it to use the same code, so that on tech done you'd also get all of the new and not-yet-mined-by-chance tiles at once...

You wanna merge this and I'll do interactivity later? Or you want it completed first? If choice 1, please wait for me to do the -n to (n) thing.

**Comment by JackRainy** [comment]
Just do (n) here. That's fair enough.
Clicking them - is just a nice to have. I can do that as well, no problem.

**Comment by rh-github-2015** [comment]
Well, I couldn't resist. I hope I have neither abused the translation system not the tutorial system with that help line that disappears forever once the 'untapped' tappability has been used... I don't like it, as I am pushing that line into a merged table row, it should go into a word-wrapping label outside, but I was too lazy to find out how. Once somebody does a really long translation - well, the table adapts and gets reeeeeaaaaalllllyyyyy wide.

Damn, that studio-eats-trailing-spaces again. Can someone tell me how to tell Android Studio to respect what I type including trailing spaces? And how to get it to not run the language file through the compiler flagging several thousand errors?

**Comment by JackRainy** [comment]
Well, saying truly, I did not look into the code deeply, however, I don't like several things already by now:
1) We have quite good Tutorial framework with popup windows (see https://github.com/yairm210/Unciv/blob/master/android/assets/jsons/Tutorials.json for the content). It is much better to use it rather than big distracting red text (is this functionality _so much_ important to tell about it like _this_?)
Screenshot 2020-03-20 20 03 51

2) Translations in `English.properties` such as: `Untapped = Unimproved` just confuse the developer who are trying to find "Unimproved" in the code and fail, because it is "Untapped" !

3) Spawning tons of the notifications about revealed resources - that`s not what I as a player would like to see:
Screenshot 2020-03-20 20 02 04
Why do we need that at all?


P.S> The grey (n) numbers look quite good. Thank you.

**Comment by rh-github-2015** [comment]
I know I know. I didn't look at how the tutorial works elsewhere. Thing is, it's just a nice helper, nobody needs this to win, so I thought a tutorial might not be the place. That red monster - I said it's not the way this should be done, right? Only if we leave it as is most will only notice what a click there does by accident. Changing all the filled cells to buttons would be too obtrusive, too.

And getting tons of the notifications - well if you leave too many resources unimproved you'll get tons. I never had more than four or five. If you mean that each cycle through the overview screen will add without cleaning up - that commented line was an attempt to do so, but it demanded an API level assertion and I don't want to impose one.

Changing the displayed word via English.properties - sorry I forgot to comment. I meant to propose a democratic choice as I never was perfectly happy with "Untapped". And to change it this way was also an experiment to see if some unit test would complain. 
If you all agree "Unimproved" is better, I'll change the key to be "Unimproved" and all will be well. Or maybe you like "Untapped"...

So - the red text gets dropped, we decide on a wording and it's ripe? Or would you prefer I cut the notification system as well?

**Comment by JackRainy** [comment]
> So - the red text gets dropped, we decide on a wording and it's ripe? Or would you prefer I cut the notification system as well?

Since I am not an owner of the game, but just a player and contributor, these questions are to @yairm210 . My role as a player to suggest UX, usability improvements, and as a developer - improve the eco-system and the quality of the code. That's not me who makes such decisions, sorry :)

**Comment by yairm210** [comment]
Okay, we're on 12 comments here already, and this was done pre-tutorial change which also doesn't help much.
I'll tell you what.
Let's start with ONLY having the 'Unimproved' lines. No click functionality, nothing, just saying to the player 'you know you have unimproved resources somewhere? Start looking, boyo'
I think that's something we can all agree on
AFTER THAT, we can discuss how this might be improved - clickable? With explanations? (I think not, BTW)
But let's start with the minimal helpful info, which is 'unimproved'

Because of the translations and whatnot, I suggest (and kinda enforce because I've been doing translation fixes for a couple of days now) that this be done in a separate PR, which is branched off of the current master

**Pull request closed without merging**